src/main/scala/linting/rule/LintConflictingModuleNames.scala

```diff
@@ -0,0 +1,314 @@
+  */
+sealed trait NamingStrategy {
+
+  /** Generates a new stable module name based on the desired name and the module IR node
+    *
+    * called by [[LintConflictingModuleNames]] to rename modules
+    *
+    * @param desiredName the requested name for the module, generated name should contain this name
+    * @param module the module targeted by the desiredName
+    */
+  def getName(desiredName: String)(module: Module): String
+}
+
+/** A naming strategy that always returns the desired name
+  */
+case object ExactNamingStrategy extends NamingStrategy {
+  def getName(desiredName: String)(module: Module): String = desiredName
+}
+
+/** Specifies a naming strategy to use for a module and the modules that it collides with
+  */
+case class NamingStrategyAnnotation(
+  strategy: NamingStrategy,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] with HasSerializationHints {
+  def typeHints: Seq[Class[_]] = Seq(
+    ExactNamingStrategy.getClass
+  )
+
+  def duplicate(newTarget: IsModule): NamingStrategyAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Specifies the original desired name for a module
+  */
+case class DesiredNameAnnotation(
+  desiredName: String,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] {
+  def duplicate(newTarget: IsModule): DesiredNameAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Specifies the desired name to rename the module to
+  */
+case class OverrideDesiredNameAnnotation(
+  desiredName: String,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] {
+  def duplicate(newTarget: IsModule): OverrideDesiredNameAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Collects [[DesiredNameAnnotation]]s for [[LintConflictingModuleNames]] to lint
+  */
+case object LintConflictingModuleNamesAspect extends Aspect[RawModule] {
+  def toAnnotation(top: RawModule): AnnotationSeq = {
+    Select.collectDeep(top) {
+      case m: chisel3.Module => DesiredNameAnnotation(m.desiredName, m.toTarget)
+    }.toSeq
+  }
+}
+
+/** An example of how to customize the names of certain modules using an Aspect
+  */
+case object StabilizeNamesAspect extends Aspect[RawModule] {
+  def toAnnotation(top: RawModule): AnnotationSeq = {
+    RunFirrtlTransformAnnotation(new RenameDesiredNames) +:
+    Select.collectDeep(top) {
+      // annotating all Queues with a more descriptive desired name
+      case m: Queue[_] => Seq(
+        new OverrideDesiredNameAnnotation(s"Queue_${m.genType.getClass.getSimpleName}_entries_${m.entries}", m.toTarget)
+      )
+      case m: TLMonitor => Seq(
+        new OverrideDesiredNameAnnotation(m.desiredName, m.toTarget)
+      )
+
+      // annotating specific instances
+      case th: TestHarness =>
+        val core = th.ldut.rocketTiles.head.module.core
+        Seq(
+          OverrideDesiredNameAnnotation("Rocket_Core_0", core.toTarget),
+          NamingStrategyAnnotation(ExactNamingStrategy, core.toTarget),
```

Comment:
these are the same now that there is only one strategy. yes it still takes effect if there is an override annotation

Reply:
Is there any reason to keep the NamingStrategyAnnotation in this PR if the only strategy is Exact?

Reply:
I think it's fine to keep it there in case we do decide to add in other strategies.

Reply:
Ok, then maybe we should add a comment to explain what each of these lines is doing, since the whole point of this file is to be an example for others to follow

Reply:
this example was moved into a scaladoc on the `RenameModulesAspect` class